pypesto/hierarchical/calculator.py

```diff
@@ -0,0 +1,173 @@
+
+        dim = len(x_ids)
+        nllh = 0.0
+        snllh = np.zeros(dim)
+        s2nllh = np.zeros([dim, dim])
+
+        res = np.zeros([0])
+        sres = np.zeros([0, dim])
+
+        # set order in solver to 0
+        amici_solver.setSensitivityOrder(0)
+
+        # fill in boring values
+        x_dct = copy.deepcopy(x_dct)
+        for key, val in self.inner_problem.get_boring_pars(
+                scaled=True).items():
+            x_dct[key] = val
+
+        # fill in parameters
+        amici.parameter_mapping.fill_in_parameters(
+            edatas=edatas,
+            problem_parameters=x_dct,
+            scaled_parameters=True,
+            parameter_mapping=parameter_mapping,
+            amici_model=amici_model
+        )
+
+        # run amici simulation
+        rdatas = amici.runAmiciSimulations(
+            amici_model,
+            amici_solver,
+            edatas,
+            num_threads=min(n_threads, len(edatas)),
+        )
+
+        self._check_least_squares(sensi_order, mode, rdatas)
+
+        # check if any simulation failed
+        if any([rdata['status'] < 0.0 for rdata in rdatas]):
+            return get_error_output(amici_model, edatas, rdatas, sensi_order, mode, dim)
+
+        sim = [rdata['y'] for rdata in rdatas]
+        sigma = [rdata['sigmay'] for rdata in rdatas]
+
+        # compute optimal inner parameters
+        x_inner_opt = self.inner_solver.solve(
+            self.inner_problem, sim, sigma, scaled=True)
+
+        # nllh = self.inner_solver.calculate_obj_function(x_inner_opt)
+
+        # print(x_inner_opt)
+
+        # if sensi_order == 0:
+        #     dim = len(x_ids)
+        #     nllh = compute_nllh(self.inner_problem.data, sim, sigma)
+        #     return {
+        #         FVAL: nllh,
+        #         GRAD: np.zeros(dim),
+        #         HESS: np.zeros([dim, dim]),
+        #         RES: np.zeros([0]),
+        #         SRES: np.zeros([0, dim]),
+        #         RDATAS: rdatas
+        #     }
+
+        # fill in optimal values
+        # TODO: x_inner_opt is different for hierarchical and
+        #  qualitative approach. For now I commented the following
+        #  lines out to make qualitative approach work.
+
+        # directly writing to parameter mapping ensures that plists do not
+        # include hierarchically computed parameters
+        for mapping in parameter_mapping:
+            for key, val in mapping.map_sim_var.items():
+                if val in x_inner_opt:
+                    mapping.map_sim_var[key] = x_inner_opt[val]
+
+        # fill in parameters
+        amici.parameter_mapping.fill_in_parameters(
+            edatas=edatas,
+            problem_parameters=x_dct,
+            scaled_parameters=True,
+            parameter_mapping=parameter_mapping,
+            amici_model=amici_model
+        )
+
+        if sensi_order == 0:
+            nllh = compute_nllh(self.inner_problem.data, sim, sigma)
+            return {FVAL: nllh,
```

Comment:
note that this piece of code will screw you over if you want to use rdata for anything  such as plotting plotting (since it contains results without scalings applied)